### ISSUE TITLE
Slow query entering Edit Order screen

### ISSUE THREAD

**Issue by harleec61** [woothemes/woocommerce#53911]
### Prerequisites

- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

My live site has recently experienced a significant slowdown in the backend. I have the latest WP and Woo installed. I'm using PHP 8.2. When clicking into a customer's order, Query Monitor is showing slow queries (Automattic\W\I\D\CustomMetaDataStore->get_meta_keys()) ranging 4-11s. When proceeding to create a label, usually in 2-3 orders at one time, the Cloudways CPU monitor shows a huge spike in utilization and the backend gets extremely slow. I created a staging site and deactivated every plugin except for Woocommerce and Query Monitor. The slow query is still highlighted. I've seen at least one other ticket describing this issue, but they appear to be closed/resolved in previous Woo releases. Is this still a known issue, or do I have a new symptom?

![Image](https://github.com/user-attachments/assets/c761f3e5-a79b-4b7e-8674-bfd810135317)
![Image](https://github.com/user-attachments/assets/4152dabb-46da-499a-bca5-1045d1d8b5c1)

### Expected behavior

I shouldn't see a slow query report in Query Monitor for a Woo query: Automattic\W\I\D\CustomMetaDataStore->get_meta_keys()

### Actual behavior

I'm seeing a slow query report in Query Monitor every time I click into the Edit Order screen.

### Steps to reproduce

I described the steps in the "Describe the bug" section above.

### WordPress Environment

<details><summary>System Status Report</summary>

```
### WordPress Environment ###

WordPress address (URL): [Redacted]
Site address (URL): [Redacted]
WC Version: 9.5.1
Legacy REST API Package Version: The Legacy REST API plugin is not installed on this site.
Action Scheduler Version: ✔ 3.8.2
Log Directory Writable: ✔
WP Version: 6.7.1
WP Multisite: –
WP Memory Limit: 1 GB
WP Debug Mode: –
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.62 (Debian)
PHP Version: 8.2.26
PHP Post Max Size: 400 MB
PHP Time Limit: 300
PHP Max Input Vars: 10000
cURL Version: 7.74.0
OpenSSL/1.1.1w

SUHOSIN Installed: –
MySQL Version: 10.6.20-MariaDB-deb11-log
Max Upload Size: 400 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

[REDACTED]

### Post Type Counts ###

acf-field: 2
acf-field-group: 1
attachment: 4317
atum_supplier: 18
aw_workflow: 1
bis_masterlog: 9
cartflows_flow: 1
cartflows_step: 3
custom_css: 5
ecwd_calendar: 1
elementor_library: 2
is_search_form: 1
lord_vaper_promos: 20
mts_notification_bar: 1
nav_menu_item: 481
oembed_cache: 63
page: 37
pewc_field: 6
pewc_group: 1
pewc_product_extra: 240
popup: 1
popup_theme: 6
post: 4
product: 1138
product_variation: 783
racemailtemplate: 1
revision: 817
seedprod: 2
shop_coupon: 2515
shop_order: 65784
shop_order_refund: 82
slider: 2
warranty_request: 1
wfacp_checkout: 7
wp_global_styles: 1
wpblogmanager: 1
wpcf7_contact_form: 1
wpcode: 2
xlwcty_thankyou: 1
zn_pb_templates: 2
znpb_template_mngr: 2

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (2) ###

Query Monitor: by John Blackbourn – 3.17.0
WooCommerce: by Automattic – 9.5.1

### Inactive Plugins (67) ###

Admin Menu Editor: by Janis Elsts – 1.12.4
Advanced Bootstrap Blocks: by  – 1.8
Advanced Coupons for WooCommerce Free: by Rymera Web Co – 4.6.4
Advanced Coupons for WooCommerce Premium: by Rymera Web Co – 4.0.1
Advanced Custom Fields PRO: by WP Engine – 6.3.11
Age Gate: by Phil Baker – 3.5.2
Ajax Search Pro: by Ernest Marcinko – 4.26.4
All-in-One WP Migration: by ServMask – 7.87
ATUM Export PRO: by Stock Management Labs™ – 1.6.3.1
ATUM Inventory Management for WooCommerce: by Stock Management Labs™ – 1.9.43.1
Breeze: by Cloudways – 2.2.0
Brevo - WooCommerce Email Marketing: by Brevo – 4.0.32
Classic Widgets: by WordPress Contributors – 0.3
Cloudflare: by Cloudflare
Inc. – 4.12.8

Coming Soon Page, Maintenance Mode, Landing Pages & WordPress Website Builder by SeedProd: by SeedProd – 6.18.12
Contact Form 7: by Takayuki Miyoshi – 6.0.1
Contact Form DB: by Michael Simpson – 2.10.36
Customer Reviews for WooCommerce: by CusRev – 5.64.1
Customer Reviews for WooCommerce PRO: by CusRev – 1.11.1
Custom Product Tabs for WooCommerce: by Code Parrots – 1.8.5
Duck Diver Product Gallery Slider for WooCommerce: by thehowarde – 3.1.4
Duck Kint Debugger: by Brian Fegter
Chris Dillon – 2.0.1

Email Customizer for WooCommerce (Pro): by ThemeHigh – 3.9.2
Freshdesk Official: by Freshworks – 2.4.1
Google Analytics for WooCommerce: by WooCommerce – 2.1.8
Health Check & Troubleshooting: by The WordPress.org community – 1.7.1
Jetpack: by Automattic – 14.1
MalCare WordPress Security Plugin - Malware Scanner, Cleaner, Security Firewall: by MalCare Security – 5.88
Max Mega Menu: by megamenu.com – 3.4.1
Multi-Carrier Shipping Plugin for WooCommerce: by PluginHive – 3.0.7
Newsletter, SMTP, Email marketing and Subscribe forms by Brevo: by Brevo – 3.1.90
NextMove Lite - Thank You Page for WooCommerce: by XLPlugins – 2.19.0
Object Cache Pro: by Rhubarb Group – 1.21.3
reCaptcha for WooCommerce: by I Thirteen Web Solution – 2.54
Regenerate Thumbnails: by Alex Mills (Viper007Bond) – 3.1.6
Route App: by Route – 2.2.22
Solid Affiliate: by Solid Plugins – 2.1.2
Solid Mail: by SolidWP – 2.1.2
SUMO Reward Points: by Fantastic Plugins – 30.0.0
Variation Swatches for WooCommerce: by RadiusTheme – 2.3.17
Variation Swatches for WooCommerce Pro: by RadiusTheme – 2.2.6
Waitlist Woocommerce Premium ( Back in stock notifier ): by  – 3.0
WooCommerce Accept.Blue: by RyanPlugins – 1.20.20
WooCommerce Anti Fraud: by OPMC Australia Pty Ltd – 6.0.2 (update to version 6.0.3 is available)
WooCommerce BlueCheck Age Verification: by BlueCheck – 4.1.1
WooCommerce Conditional Shipping and Payments: by Woo – 2.0.4
WooCommerce Freshdesk: by Themesquad – 1.3.0 (update to version 1.3.1 is available)
WooCommerce Print Invoices/Packing Lists: by SkyVerge – 3.13.7
WooCommerce Product Add-Ons Ultimate: by Plugin Republic – 3.22.0
WooCommerce Sequential Order Numbers Pro: by SkyVerge – 1.21.3
WooCommerce Shipment Tracking: by WooCommerce – 2.5.2 (update to version 2.5.3 is available)
WooCommerce Shipping: by WooCommerce – 1.3.1
WooCommerce Shipping & Tax: by WooCommerce – 2.8.5
WooCommerce Show Single Variations by Iconic: by Iconic – 1.18.0
WooCommerce Thank You Page - NextMove: by XLPlugins – 1.17.1
WP-Sweep: by Lester 'GaMerZ' Chan – 1.1.8
WP Blog Manager: by AccessPress Themes – 2.0.1
WPCode Lite: by WPCode – 2.2.4.1
WPC Order Notes for WooCommerce: by WPClever – 1.5.3
WP Crontrol: by John Blackbourn – 1.17.1
YITH WooCommerce Frequently Bought Together Premium: by YITH – 1.46.0
Yoast SEO: by Team Yoast – 24.0
Yoast SEO: WooCommerce: by Team Yoast – 16.4
Yoast SEO Premium: by Team Yoast – 24.0
YouTube WordPress Plugin by Embed Plus: by Embed Plus YouTube Plugin Team – 14.2.1.2
Zion Builder: by zionbuilder.io – 3.6.12
Zion Builder Pro: by zionbuilder.io – 3.6.12

### Dropin Plugins () ###

db.php: Query Monitor Database Class (Drop-in)

### Must Use Plugins (1) ###

Health Check Troubleshooting Mode: by  – 1.9.2

### Settings ###

Legacy API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-filtered (exclude-from-filtered)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: ✔
Enforce Approved Product Download Directories: –
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: –

### Logging ###

Enabled: ✔
Handler: Automattic\WooCommerce\Internal\Admin\Logging\LogHandlerFileV2
Retention period: 30 days
Level threshold: –
Log directory size: 4 MB

### WC Pages ###

Shop base: #592 - /shop/
Cart: #593 - /cart/ -  Contains the [woocommerce_cart] shortcode
Checkout: #594 - /checkout/ -  Contains the [woocommerce_checkout] shortcode
My account: #595 - /my-account/
Terms and conditions: #868 - /terms-conditions/

### Theme ###

Name: Lord Vaper Pens 2022
Version: 1.0
Author URL: https://www.duckdiverllc.com
Child Theme: ✔
Parent Theme Name: DuckDiverFramework
Parent Theme Version: 1.2.29
Parent Theme Author URL: https://www.howardehrenberg.com
Theme type: Classic theme
WooCommerce Support: ✔

### Templates ###

Overrides: LV_Theme_22/woocommerce/archive-product.php version 3.4.0 is out of date. The core version is 8.6.0
LV_Theme_22/woocommerce/cart/cart-totals.php
LV_Theme_22/woocommerce/cart/cart.php

Outdated Templates: ❌
					
					
						Learn how to update					 |
					
						
					
					
						Clear system status theme info cache

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customize-store
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
onboarding
onboarding-tasks
pattern-toolkit-full-composability
product-custom-fields
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
printful
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page
launch-your-store

Disabled Features: product-data-views
experimental-blocks
coming-soon-newsletter-template
minified-js
product-pre-publish-modal
settings
async-product-editor-category-field
product-editor-template-system
blueprint
reactify-classic-payments-settings
use-wp-horizon
add-to-cart-with-options-stepper-layout

Daily Cron: ✔ Next scheduled: 2024-12-20 09:02:33 -08:00
Options: ✔
Notes: 22
Onboarding: completed

### Action Scheduler ###

Complete: 60,398
Oldest: 2024-11-18 23:26:11 -0800
Newest: 2024-12-19 22:31:36 -0800

Failed: 1,110
Oldest: 2021-09-14 17:00:53 -0700
Newest: 2024-12-17 08:33:16 -0800

Pending: 8
Oldest: 2024-12-20 08:21:25 -0800
Newest: 2025-01-20 21:21:42 -0800


### Status report information ###

Generated at: 2024-12-19 23:20:28 -08:00
```
</details>

### Isolating the problem

- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**Comment by DuckDivers** [comment]
I have looked at this with @harleec61 

There appears to be a huge performance hit in the ```Automattic\W\I\D\CustomMetaDataStore->get_meta_keys()```

In the change from Version 8.8.0 from this commit here https://github.com/woocommerce/woocommerce/commit/63443b502d9236eb1e2ef20915d0bf0ef298af02

Where the addition of this clause 

```$query .= "WHERE meta_key != '' ";```

Would apparently change the query (on his server) take from 2.9 seconds (still slow) to 15 seconds or more.  

Testing on another server the same goes from 3.5 seconds on latest WC Version down to 1.7 seconds on 8.8.0 (without the new clause).

**Comment by harleec61** [comment]
Greetings Woo team! I hope this ticket will get attention soon as this issue is seriously impacting all backend tasks. Processing and shipping orders is taking 2-3x longer due to the significant DB query delays and the spike in CPU utilization that brings the backend to a screeching halt.

**Comment by harleec61** [comment]
How long does it take to get a response? My website is having serious problems when processing orders in the backend when the DB is being heavily accessed. After extensive debugging, going back to V8.8.0 resolves it, so it's clear there's something with Woo that's affecting performance.

**Comment by DuckDivers** [comment]
So, in the case for @harleec61 - he is not using the custom field dropdown (anywhere on his site), and since the WC function that causes the slow query is triggered by https://github.com/woocommerce/woocommerce/blob/1a5e68311c8065d3cb00b6634135a2631d569909/plugins/woocommerce/src/Internal/Admin/Orders/MetaBoxes/CustomMetaBox.php#L126

Which defines `$keys` by this: 
https://github.com/woocommerce/woocommerce/blob/1a5e68311c8065d3cb00b6634135a2631d569909/plugins/woocommerce/src/Internal/Admin/Orders/MetaBoxes/CustomMetaBox.php#L115-L116

Using the WP Filter `postmeta_form_keys` and returning a simple array eliminates the query all together on the orders screen, and it cuts the page load time to a reasonable amount. 

Perhaps a simple solution is to allow people to toggle off the option at all to use the custom fields at all, since in the WP docs, it calls this an expensive query.

https://developer.wordpress.org/reference/hooks/postmeta_form_keys/#description

**Comment by prettyboymp** [comment]
I think the best solution here is to switch to loading the values of the meta key options via an asynchronous request combined with caching the values.  This option will:

* Allow the order edit screen to render without the need to wait for the query to run.
* Introduce caching that will reduce the number of times the expensive query needs to run overall.
* Minimize the impact to users who currently use the Custom Fields meta box as part of their existing workflow.
* Avoid introducing new settings/options.

### Suggested Implementation Behavior

* The meta key drop down will initially render as an empty placeholder and indicate that the options are loading.
  * Note: the dropdown shouldn't be rendered at all and the ajax if `! current_user_can( 'edit_others_shop_orders' )` as it will always be empty.
* On page load, JS will be used to send a request through `admin-ajax.php` to retrieve the list of meta keys.
* Once the list of meta keys is returned from the ajax call:
  * The options in the select box should be replaced with the returned values, prepended with the current `#NONE/Select` default.
  * A button will appear or be enabled near the drop down with the option to "refresh" the list.  This button will re-trigger the ajax request, but with an additional argument to refresh the cached list.
* The ability to switch to the "Enter new" interface should continue work as it does now, allowing users to add their own meta even if it doesn't exist in the list of current meta keys.
  * Note: One exception to this is that the new behavior will now still show the select box even if there are no existing meta keys to render in order to show the option to refresh cache.
* The server side handler of the ajax request will: 
  * Maintain the same filter behavior with the use of the `postmeta_form_keys` as it exists in
https://github.com/woocommerce/woocommerce/blob/1a5e68311c8065d3cb00b6634135a2631d569909/plugins/woocommerce/src/Internal/Admin/Orders/MetaBoxes/CustomMetaBox.php#L115-L116
  * Introduce caching around the call to retrieve the meta keys from the Datastore.
    * The cache should be indefinite (without expiration).
    * The ajax handler should support an option to refresh the cached list.
  * Reduce the resulting list of meta keys by the same checks currently implemented:
https://github.com/woocommerce/woocommerce/blob/1a5e68311c8065d3cb00b6634135a2631d569909/plugins/woocommerce/src/Internal/Admin/Orders/MetaBoxes/CustomMetaBox.php#L165

***

@harleec61 , in the meantime, if you need a simple solution you can add the following snippet to your theme's `functions.php` or similar:

```php
add_filter( 'postmeta_form_keys', '__return_empty_array' );
```

**Comment by DavidAnderson684** [comment]
This was previously reported in https://github.com/woocommerce/woocommerce/issues/47074 and there was marked as resolved by @jorgeatorres in fixes in 8.9.0 and 9.0.0, which is evidently not the case  . I'm also experiencing it (in 9.3.4 - code looks unchanged in 9.5.2).
(There's a work-around above, of course).
In my case, the slow query was 16 seconds.
